packages/core/src/common/contribution-filter/contribution-filter.ts

```diff
@@ -0,0 +1,40 @@
+/********************************************************************************
+ * Copyright (C) 2021 STMicroelectronics and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { interfaces } from 'inversify';
+import { Filter } from './filter';
+import { NameBasedFilter } from './string-based-filter';
+export type ContributionType = interfaces.ServiceIdentifier<unknown>;
+
+export const ContributionFilter = Symbol('ContributionFilter');
+
+/**
+ * Specialized `Filter` that is used by the `ContainerBasedContributionProvider` to
+ * filter unwanted contributions that are already bound in the DI container.
+ */
+export interface ContributionFilter extends Filter<Object> {
+    /**
+     * Contribution types for which this filter is applicable. If `undefined` or empty this filter
+     * will be applied to all contribution types.
+     */
+    contributions?: ContributionType[];
```

Comment:
The contribution type should not be part of the `ContributionFilter` interface: it makes it impossible to reuse the same filter implementation for multiple contributions. We can just remove this interface and rewrite `FilterRegistry.addFilter` like so:
```
addFilter(filter: Filter, restrictTo?: ContributionType[]);
```` 
if you don't pass `restrictTo`, it applies to all contributions. More philosophically, it does not feel to me that the contribution type is part of the state of the Filter, it's part of the state of the registry. On indication of that is that once the filter is added to the registry, the "contributions" field becomes unused.
I think the base problem is that we're missing a "ContributionFilterContribution" type.

Reply:
You can very well reuse filters right now: the `ContributionFilter` type is just a `Filter<Object>` with that `contributions` field. So if you want to reuse a filter, you would implement `Filter<Object>` and subclass it as many times as you want in various `ContributionFilter` implementations. I think this pattern is "forced" upon us to work with Inversify... It's not so much that `contributions` is really part of the state of the contribution filter, but rather it is used to communicate with the registry through Inversify injection.

Ultimately the issue is that `ContributionFilter` are injectable classes. This makes it hard to pass around a `Filter` constructor to resolve without the list of contributions it applies to.

The `ContributionFilterContribution` could be a different API indeed.

Reply:
@tsmaeder here's an attempt at making it more functional: https://github.com/eclipsesource/theia/compare/filter-api...eclipse-theia:mp/filter-api-4

See [sample-filtered-command-contribution.ts](https://github.com/eclipse-theia/theia/blob/7d860c78446780de7e04465ef9c387c24db7ddab/examples/api-samples/src/browser/contribution-filter/sample-filtered-command-contribution.ts#L48-L65) for usage example. I had to use a different kind of contribution mechanism, one closer to what we currently do for commands.

Reply:
I'm in favor of a more functional approach and would like to squash the proposal from @marechal-p.
@tsmaeder Do you have any objections or additional input?